File: CodeH-Java/src/main/java/com/codeh/thread/Application.java
package com.codeh.thread;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.concurrent.*;

/**
 * @author jinhua.xu
 * @version 1.0
 * @className Application
 * @date 2021/12/13 9:44
 * @description TODO
 */
public class Application {
    private static final Integer MAX_RETRIES = 3;
    private static ExecutorService executorService = Executors.newFixedThreadPool(10);
    private static Object lock = new Object();
    private static volatile int count = 0;

    public static void main(String[] args) throws Exception {
        HashMap<TaskMetaData, Future<Integer>> map = new HashMap<>();
        for (int i = 0; i < 10; i++) {
            TaskMetaData taskMetaData = new TaskMetaData(i, "task-" + i, 0);
            map.put(taskMetaData, submit(0));
        }

        boolean flag = true;
        while (flag) {
            ArrayList<TaskMetaData> keys = new ArrayList<>();
            for (TaskMetaData taskMetaData : map.keySet()) {
                keys.add(taskMetaData);
            }

            for (TaskMetaData key: keys) {
                Future<Integer> future = map.get(key);
                System.out.println(key);
                if (future.isDone()) {
                    try {
                        System.out.println(String.format("Done Thread: %s", key.getTaskName()));
                        // 获取结果，如果抛出异常，说明任务运行有问题，需要进行重试
                        future.get();
                        // 任务正常运行完成，移除这个任务
                        map.remove(key);
                    } catch (ExecutionException e) {
                        // 对异常任务进行重试
                        e.printStackTrace();
                        System.out.println(String.format("ReStart submit task: %s", key.getTaskName()));
                        if (key.getRetries() > MAX_RETRIES) {
                            // 超过最大重试次数达到上限
                            throw new Exception(String.format("%s number of retries reaches the limit", key.getTaskName()));
                        } else {
                            System.out.println(String.format("线程%s进行第%d次重试", key.getTaskId(), key.getRetries() + 1));
                            key.setRetries(key.getRetries() + 1);
                            Future<Integer> retryFuture = submit(2);
                            map.put(key, retryFuture);
                        }
                    }
                }

                if (map.isEmpty()) {
                    flag = false;
                }
            }
        }

        executorService.shutdown();
    }

    public static Future<Integer> submit(int sleepTime) throws InterruptedException {
        TimeUnit.SECONDS.sleep(sleepTime);
        Future<Integer> future = executorService.submit(new Callable<Integer>() {
            @Override
            public Integer call() throws Exception {
                for (; ; ) {
                    synchronized (lock) {
                        count++;
                        System.out.println(String.format("Current thread: %s count: %d", Thread.currentThread(), count));
                        if (count == 50 || count == 100) {
                            System.out.println(String.format("[Error] thread: %s", Thread.currentThread()));
                            throw new Exception(String.format("count: %d", count));
                        }

                        if (count > 200) {
                            break;
                        }
                    }
                }
                return 0;
            }
        });

        return future;
    }
}

@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
class TaskMetaData{
    private Integer taskId;
    private String taskName;
    private Integer retries;
}

File: CodeH-Scala/src/main/scala/com/codeh/thread/Application.scala
package com.codeh.thread

import java.util.concurrent.{Callable, ExecutorService, Executors, Future, TimeUnit}

import scala.collection.mutable
import scala.collection.mutable.ListBuffer
import util.control.Breaks._

/**
 * @className Application
 * @author jinhua.xu
 * @date 2021/12/13 10:45
 * @description 线程池提交任务并发处理，并对线程级别的重试处理
 * @version 1.0
 */
object Application {
  val executorService: ExecutorService = Executors.newFixedThreadPool(10)
  val max_retries: Int = 3
  @volatile var count: Int = 0
  val lock: Object = new Object

  def main(args: Array[String]): Unit = {
    // 创建一个可变map容器，用于存放返回的future
    val map: mutable.Map[TaskMetaData, Future[Int]] = mutable.HashMap[TaskMetaData, Future[Int]]()
    for (i <- 1 to 10) {
      val taskMetaData: TaskMetaData = new TaskMetaData(i, "task-" + i, 0)
      map.put(taskMetaData, submit(2))
    }

    TimeUnit.SECONDS.sleep(2)

    var flag: Boolean = true
    while (flag) {
      val keys: mutable.ListBuffer[TaskMetaData] = ListBuffer[TaskMetaData]()
      for (taskMetaData <- map.keys) {
        keys += taskMetaData
      }

      for (key: TaskMetaData <- keys) {
        val future: Future[Int] = map.get(key).get
        println("task_id: " + key.taskId + " task_name: " + key.taskName + " retries: " + key.retries)
        if (future.isDone) {
          try {
            println(String.format("Done Thread:%s", key.taskName))
            future.get()
            map -= key
          } catch {
            case _: Exception => {
              println(String.format("start resubmit task: %s", key.taskName))
              if (key.retries > max_retries) {
                throw new Exception("test")
              } else {
                key.retries += 1
                println("task_id: " + key.taskId + " task_name: " + key.taskName + " retries: " + key.retries)
                val retryFuture: Future[Int] = submit(2)
                map.put(key, retryFuture)
              }
            }
          }
        }

        if (map.isEmpty) {
          flag = false
        }
      }
    }

    executorService.shutdown()
  }

  def submit(sleepTime: Int): Future[Int] = {
    TimeUnit.SECONDS.sleep(sleepTime)
    val result: Future[Int] = executorService.submit(new Callable[Int] {
      override def call(): Int = {
        for (num <- 1 to 100){
          lock.synchronized {
            breakable {
              count += 1
              println("current thread name " + Thread.currentThread().getName + " count: " + count)
              //              println(String.format("current thread: %s count: %d", Thread.currentThread(), count))
              if (count == 50 || count == 100) {
                println(String.format("[Error] thread: %s", Thread.currentThread().getName))
                //                throw new Exception(String.format("count: %d", count))
                throw new Exception("fail thread name is " + Thread.currentThread().getName)
              }
              if (count > 200) {
                break()
              }
            }
          }
        }
        0
      }
    })

    result
  }
}

class TaskMetaData(task_id: Int, task_name: String, inRetry: Int) {
  var taskId: Int = task_id
  var taskName: String = task_name
  var retries: Int = inRetry
}
